File: src/libraries.py
"""
@brief This file contains library lookup and cache generation.
@info SB has two types of caches pertaining to libraries. The first
is the lib.cache stored for each application that contains both individual .so
libraries, and also folders, within the /usr/lib directory (Which is symlinked to
/lib). This cache contains all the contents of /usr/lib that the application needs
to run, and is determined by recursively calling ldd on all binaries the application
needs. This can be updated with --update-libraries, and is a fast operation
@info The second cache is the folder cache that determines libraries OUTSIDE of a
folder in /usr/lib is needed for those libraries within it; for example:
ldd /usr/lib/qt6/bin/qtpaths will return a dependency on /usr/lib/libcap.so.2
We don't want to hardlink each file in the directory, so we instead bind mount
the whole folder, and add any individual libraries that the folder might need. As
the example suggests, this is entirely due to the /usr/lib/qt{5,6,} folders. Unlike
--update-libraries, the corresponder updator --update-library-cache is very slow, hence
why we've created a second cache to abstract from the main application cache. Now, applications
can simply read a cache file containing all the libraries needed for a certain folder. sb-refresh
will delete both caches.
"""

from pathlib import Path
from concurrent.futures import ThreadPoolExecutor, as_completed

from shared import args, output, log, data


# The set of libraries the application currently needs.
current = set()


# Used for library dependency lookup. Libraries will contain all the needed libraries,
# Searched will contain libraries that have already been searched.
searched = {""}


# A set of wildcard libraries
wildcards = set()


def parse_ldd(to_load, recursive=True):
    ret = {to_load}
    for library in output(["ldd", to_load]):
        split = library.split()

        # Handle the esoteric syntax of ldd.
        lib = None
        if len(split) > 0 and split[0].startswith("/usr/lib/"):
            lib = split[0]
        elif len(split) > 2 and split[2] != "not":
            lib = split[2]
            lib = lib.replace("lib64", "lib")
            if lib.startswith("/lib"):
                lib = lib[4:]
        if lib is not None:
            ret |= get(lib) if recursive else {lib}
    return ret


def get(to_load):
    """
    @brief Get the current libraries for a binary.
    @param to_load: The binary or library
    @returns The libraries needed for the binary
    """

    global searched, wildcards

    ret = set()

    # If the binary/library has already been searched, just return, otherwise add it.
    if to_load in searched or to_load == "":
        return ret
    searched.add(to_load)

    # If its a directory, then use the directory cache.
    if Path(to_load).is_dir():
        cache = Path(data, "sb", "cache")
        cache.mkdir(parents=True, exist_ok=True)

        sub = to_load.replace("/", ".")
        if not sub.endswith("."):
            sub += "."
        sub += "cache"

        dir_cache = cache / sub

        if not dir_cache.is_file() or args.update_cache:

            libs = output(["find", str(to_load), "-mindepth", "1", "-executable", "-type", "f"])
            running = set()

            with ThreadPoolExecutor() as executor:
                futures = {executor.submit(parse_ldd, lib): lib for lib in libs}
                for future in as_completed(futures):
                    result = future.result()
                    running |= result

            internal = set()
            for lib in running:
                if lib.startswith(str(to_load) + "/"):
                    internal.add(lib)

            running = list(running - internal)
            if not running:
                return ret
            with dir_cache.open("w") as file:
                for library in running[:-1]:
                    if library != "" and library != "not":
                        file.write(f"{library} ")
                file.write(running[-1])

        if dir_cache.is_file():
            libs = set(dir_cache.open("r").readline().strip().split(" "))
            ret |= libs
            ret.add(to_load)
        return ret

    # If there's a wildcard, expand it.
    if "*" in to_load:
        wildcards.add(to_load)
        return ret

    # Otherwise, add the library/binary to libraries, and then add all libraries needed via ldd.
    ret |= parse_ldd(to_load)
    return ret


def setup(sof_dir, lib_cache, update_sof):
    """
    @brief Setup the library folder.
    @param sof_dir: Where the SOF dir is located.
    @param lib_cache: Where the library cache is.
    @param update_sof: Whether we should actually update the SOF if it is already present.
    """

    global current
    binds = set()

    # Make the directory if it doesn't exist.
    if not sof_dir.is_dir():
        sof_dir.mkdir(parents=True, exist_ok=True)

    # If we want to update the SOF, perform the recursive library dependency check.
    if update_sof:

        log("Finding wildcard libraries...")
        command = ["find", "/usr/lib", "-mindepth", "1", "-maxdepth", "1", "-executable"]
        for library in wildcards:
            command.extend(["-name", library[library.rfind("/") + 1:], "-o"])
        # Remove the trailing -o
        command = command[:-1]
        current |= set(output(command))

        # We simply get libraries based on everything within libraries that hasn't been searched.
        # Run this however many times it takes before all dependencies are found.
        log("Determining library dependencies...")
        unsearched = current - searched
        with ThreadPoolExecutor() as executor:
            while unsearched:
                futures = {executor.submit(get, lib): lib for lib in unsearched}
                for future in as_completed(futures):
                    current |= future.result()
                unsearched = current - searched

    # When updating, we need to overwrite the lib_cache.
    with lib_cache.open("w") as cache:
        for lib in current:
            cache.write(lib)
            cache.write(" ")

    # Now, create our library folder, by copying everything in libraries into the folder.
    # We use a SHARED folder, and then create hard-links off of that, to reduce duplication
    # between multiple applications running under SB.
    log("Creating /lib...")

    # Create the shared runtime.
    runtime_lib = Path("/tmp", "sb", "shared")
    runtime_lib.mkdir(parents=True, exist_ok=True)

    # Create the application-specific folder.
    sof_dir.mkdir(parents=True, exist_ok=True)

    # For each library, add it to the shared runtime.
    # Threading does not lead to noticable results, unfortunately.
    for library in current:

        # Fix some issues that sometimes happen with find.
        if not library.startswith("/usr/lib") and not library.startswith("/usr/bin"):
            log(f"Ignoring invalid library: {library}")
            continue

        # Get the path to the library, and where it should end up in the shared runtime.
        real_path = Path(library)
        runtime_path = Path(f"{runtime_lib}/{library}")

        # If the library isn't a directory, then write it.
        if not real_path.is_dir():
            write(library, runtime_path, real_path, sof_dir)

        # If it IS a directory, then return it for binding.
        elif real_path.is_dir():
            binds.add(str(real_path))
    return binds


def write(library, runtime_path, real_path, sof_dir):
    """
    @brief Write a file to the SOF library folder.
    @param library: The name of the library.
    @param runtime_path: Where the shared library folder is.
    @param real_path: Where the path is on the actual system.
    @param sof_dir: The path to the program's SOF.
    """

    # If the library isn't in the shared folder, then add it.
    if not runtime_path.is_file() and str(real_path).startswith("/usr/lib/"):

        # Make any needed directories.
        runtime_path.parents[0].mkdir(parents=True, exist_ok=True)

        # Try to read the library, and dump it into the runtime.
        # Then, set the correct CHMOD.
        try:
            runtime_path.open("wb").write(real_path.open("rb").read())
            runtime_path.chmod(real_path.stat().st_mode)

        # Handle exceptions, but we don't really care.
        except PermissionError:
            log("Permission Denied: ", str(real_path))
            return
        except FileNotFoundError:
            log("File not found:", str(real_path), "This is fine")
            return

    # If we are dealing with files, not folders, Then hardlink the shared library into the SOF.
    if runtime_path.is_file() and real_path.is_file():
        lib_path = Path(f"{str(sof_dir)}/{library}")
        if runtime_path.is_file() and not lib_path.is_file():
            try:
                lib_path.parents[0].mkdir(parents=True, exist_ok=True)
                lib_path.hardlink_to(runtime_path)
                lib_path.chmod(real_path.stat().st_mode)
            except Exception:
                pass

File: cpp/src/libraries.cpp
#include <algorithm>
#include <filesystem>
#include <fstream>
#include <regex>
#include <filesystem>

#include "shared.hpp"
#include "libraries.hpp"
#include "arguments.hpp"

using namespace shared;

namespace libraries {

  // Directories, so we can mount them after discovering dependencies.
  lib_t directories = {};

  // Get the name for a cache.
  inline std::filesystem::path cache_name(std::string library, const std::string& function) {
    std::replace(library.begin(), library.end(), '/', '.');
    library = strip<char>(library, '*');
    auto cache = std::filesystem::path(data) / "sb" / "cache" / std::string(library + '.' + function + ".lib.cache");
    return cache;
  }

  void parse_ldd(lib_t& required, const std::string& library, const std::string_view& directory) {
    lib_t libraries;

    if (!std::filesystem::exists(library)) return;

    auto cache = cache_name(std::string(library), "ldd");
    if (std::filesystem::exists(cache) && !std::filesystem::is_empty(cache))
      libraries = read_file<set>(cache, setorize);
    else {
      if (read_file<std::string>(library, head<5>) != "\177ELF\2") return;
      std::filesystem::create_directories(cache.parent_path());

      // If this is a library, add it.
      if (std::filesystem::exists(library) && library.starts_with("/usr/lib/") && directory == "") {
        libraries.emplace(library);
      }

      // LDD it!
      auto output = exec<std::string>({"ldd", library}, dump, STDOUT);

      // Parse it in a single pass, rather than splitting.
      for (size_t x = output.find('/'); x != std::string::npos; x = output.find('/', x + 1)) {
        auto end = output.find(' ', x);
        if (end < output.size()) {
          auto shared_lib = output.substr(x, end - x);

          // Normalize.
          shared_lib = std::regex_replace(shared_lib, std::regex("lib64"), "lib");
          if (!shared_lib.starts_with("/usr/lib/")) {
            if (shared_lib.starts_with("/lib/")) shared_lib.insert(0, "/usr");
            else if (shared_lib.starts_with("/")) shared_lib.insert(0, "/usr/lib");
          }

          // Avoid resolving libraries in the directory itself.
          if (!directory.empty()) {
            if (shared_lib.starts_with(directory)) continue;
          }

          if (shared_lib.contains("..")) shared_lib = "/usr/lib/" + shared_lib.substr(shared_lib.rfind('/') + 1);

          // Ensure it's actually here.
          if (std::filesystem::exists(shared_lib) && shared_lib.starts_with("/usr/lib")) libraries.emplace(shared_lib);
        }
      }

      if (!libraries.empty()) {
        auto file = std::ofstream(cache);
        if (file.is_open()) {
          file << join(libraries, ' ');
          file.close();
        }
        else log({"Failed to write cache file:", cache.string()});
      }
    }
    required.merge(libraries);
  }


  // Get all dependencies for a library/binary.
  void get(lib_t& libraries, const std::string_view& library, std::string directory) {
    lib_t local = {};

    // Our cached definitions
    auto cache = cache_name(std::string(library), "get");

    // If the cache exists, and we don't need to update, use it.
    if (std::filesystem::exists(cache) && !std::filesystem::is_empty(cache)) {
      libraries.merge(read_file<set>(cache, setorize));
      return;
    }

    std::filesystem::create_directories(cache.parent_path());

    // Resolve wildcards
    if (library.contains("*"))
     local.merge(wildcard(library, "/usr/lib", {"-maxdepth", "1", "-type", "f,l", "-executable"}));

    // Find all shared libraries in dir.
    else if (std::filesystem::is_directory(library)) {
      local.merge(exec<lib_t>({"find", library, "-type", "f,l", "-executable"}, fd_splitter<lib_t, '\n'>, STDOUT));
      directory = library;
    }

    // Just use the direct dependencies of the library
    else if (library.starts_with("/usr/lib")) local.emplace(library);
    
    // Binaries just exclude itself.
    else parse_ldd(libraries, library.data(), "");

    batch(parse_ldd, local, local, directory);

    if (!local.empty()) {
      auto file = std::ofstream(cache);
      if (file.is_open()) {
        file << join(local, ' ');
        file.close();
      }
      else log({"Failed to write cache file:", cache.string()});
    }
    libraries.merge(local);
  }

  // Setup the SOF
  void setup(const vector& libraries, const std::string_view& application, const std::filesystem::path& app_sof) {
    const auto share_dir = std::filesystem::path(arg::get("sof")) / "shared";
    std::filesystem::create_directories(share_dir);
    if (std::filesystem::is_directory(app_sof) && arg::at("update") < "libraries" && !std::filesystem::is_empty(app_sof)) return;
    std::filesystem::create_directories(app_sof);

    // Write Lambda. We batch writes to tremendously speed up initial SOF generation.
    auto write_library = [&share_dir, &app_sof, &libraries](const uint_fast32_t x) {
      const std::string& lib = libraries[x];

      // We only write normalized library files, directories are mounted
      if (!lib.starts_with("/usr/lib/") || std::filesystem::is_directory(lib)) return;

      const auto base = lib.substr(lib.find("/lib/") + 5);
      auto shared_path = share_dir / base;
      auto local_path = app_sof / base;

      // Files are actually written to the shared directory,
      // and then hard-linked into the respective directories.
      // This shared libraries between applications.
      if (std::filesystem::exists(local_path)) return;

      if (std::filesystem::is_symlink(lib)) {
        auto resolved = std::filesystem::read_symlink(lib);
        auto local_resolved = app_sof / resolved;

        if (!std::filesystem::exists(local_resolved)) {
          auto shared_resolved = share_dir / resolved;
          if (!std::filesystem::exists(shared_resolved)) {
            std::filesystem::copy_file(lib.substr(0, lib.rfind('/')) / resolved, shared_resolved);
          }
          std::filesystem::create_hard_link(shared_resolved, local_resolved);
        }
        std::filesystem::create_symlink(resolved, local_path);
      }

      else {
        const auto dir = std::filesystem::path(base).parent_path();
        if (!std::filesystem::exists(shared_path)) {
          std::filesystem::create_directories(share_dir / dir);
          std::filesystem::copy_file(lib, shared_path);
        }
        std::filesystem::create_directories(app_sof / dir);
        std::filesystem::create_hard_link(shared_path, local_path);
      }
    };

    auto lock_file = share_dir / "sb.lock";
    while (std::filesystem::exists(lock_file)) {
      log({"Another instance of the application is writing to the SOF. Waiting..."});
      auto wd = inotify_add_watch(inotify, lock_file.c_str(), IN_DELETE_SELF);
      inotify_wait(wd);
    }
    std::ofstream lock(lock_file);

    // Zoom.
    //for (size_t x = 0; x < vec.size(); ++x) write_library(x);
    auto futures = pool.submit_loop(0, libraries.size(), write_library);
    futures.wait();
    std::filesystem::remove(lock_file);
  }
  void setup(const set& libraries, const std::string_view& application, const std::filesystem::path& app_sof) {
    vector vec; vec.reserve(libraries.size());
    vec.insert(vec.end(), std::make_move_iterator(libraries.begin()), std::make_move_iterator(libraries.end()));
    setup(vec, application, app_sof);
  }

  // Symlink libraries
  void symlink(vector& command) {
    extend(command, {
      "--symlink", "/usr/lib", "/lib64",
      "--symlink", "/usr/lib", "/lib",
      "--symlink", "/usr/lib", "/usr/lib64",
    });
  }
}
